File: 2015/day03.cpp
#include <set>
#include <utility>

#include "../common/puzzle.hpp"

class Solution : public Puzzle {
   public:
    Solution(std::string inputFileName = "inputs/2015/3.txt")
        : Puzzle(inputFileName) {}
    std::string part1() {
        std::set<std::pair<int, int>> s;
        s.insert({0, 0});

        int x = 0, y = 0;
        char ch;
        while (inputFile >> ch) {
            switch (ch) {
                case '>':
                    x++;
                    break;
                case '<':
                    x--;
                    break;
                case '^':
                    y++;
                    break;
                case 'v':
                    y--;
                    break;
            }
            s.insert({x, y});
        }
        return std::to_string(s.size());
    }
    std::string part2() {
        std::set<std::pair<int, int>> s;
        s.insert({0, 0});

        int x = 0, y = 0, rx = 0, ry = 0;
        bool robot = true;
        char ch;
        while (inputFile >> ch) {
            if (!robot) {
                switch (ch) {
                    case '>':
                        x++;
                        break;
                    case '<':
                        x--;
                        break;
                    case '^':
                        y++;
                        break;
                    case 'v':
                        y--;
                        break;
                }
                s.insert({x, y});
            } else {
                switch (ch) {
                    case '>':
                        rx++;
                        break;
                    case '<':
                        rx--;
                        break;
                    case '^':
                        ry++;
                        break;
                    case 'v':
                        ry--;
                        break;
                }
                s.insert({rx, ry});
            }
            robot = !robot;
        }
        return std::to_string(s.size());
    }
};

int main() {
    Solution d = Solution();
    d.solve();
}

File: year2016/day03.java
package year2016;

import java.util.Scanner;

import common.puzzle;

public class day03 extends puzzle {
    public day03() {
        super("./inputs/2016/3.txt");
    }

    public static void main(String[] args) {
        new day03().solve();
    }

    protected String part1() {
        Scanner reader = getScanner();
        int counter = 0;
        while (reader.hasNextLine()) {
            String s[] = reader.nextLine().trim().split("\\s+");
            int a = Integer.parseInt(s[0]);
            int b = Integer.parseInt(s[1]);
            int c = Integer.parseInt(s[2]);
            if (a + b > c && a + c > b && c + b > a)
                counter++;
        }
        return Integer.toString(counter);
    }

    protected String part2() {
        Scanner reader = getScanner();
        int counter = 0;
        while (reader.hasNextLine()) {
            String s1[] = reader.nextLine().trim().split("\\s+");
            String s2[] = reader.nextLine().trim().split("\\s+");
            String s3[] = reader.nextLine().trim().split("\\s+");
            for (int i = 0; i < 3; i++) {
                int a = Integer.parseInt(s1[i]);
                int b = Integer.parseInt(s2[i]);
                int c = Integer.parseInt(s3[i]);
                if (a + b > c && a + c > b && c + b > a)
                    counter++;
            }
        }
        return Integer.toString(counter);
    }
}